### PR TITLE
many: do without device state/assertions accessors based on state only outside of devicestate/tests

### PR DESCRIPTION
**Pull request by pedronis**
Because under remodeling  we have two models, the current one, and the new model to be,  lots of operations to be correct need to get the device state, model assertions in a contextual manner, via the new snapstate.DeviceCtx for example. The previous accessors solely starting from state are dangerous, do without most of them and reduce the surface of DeviceManager as well.  

Device and SetDevice are still available for tests through devicestatetest (This is achieved introducing a devicestate/internal package).

DeviceManager is not directly a storecontext.Backend anymore, but can provide one via its StroreContextBackend method.

This leaves a few XXXs that will need to be addressed soon.

A follow will also use a registrationContext indirection instead of directly the DeviceManager in much of the registration handling code, to allow its reuse in a remodeling change.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -2125,12 +2125,13 @@ func getUserDetailsFromStore(theStore snapstate.StoreService, email string) (str
 	return v.Username, opts, nil
 }
 
-func createAllKnownSystemUsers(st *state.State, createData *postUserCreateData) Response {
+func createAllKnownSystemUsers(o *overlord.Overlord, createData *postUserCreateData) Response {
```
Comment (stolowski) [inline daemon/api.go]: I'm not too fond of passing Overlord here just to call Model() and State() on it (but I might be missing something and haven't looked at remaining PRs yet) as IMHO it creates tight coupling with overlord and makes testing potentially harder, and also makes the intent of the arguments less obvious. How about passing state as before, and having a Modeler (ModelGetter) interface  argument instead (the interface would declare just the Model() method)?

File: daemon/api.go [hunk 2]
```diff
@@ -2179,12 +2180,13 @@ func createAllKnownSystemUsers(st *state.State, createData *postUserCreateData)
 	return SyncResponse(createdUsers, nil)
 }
 
-func getUserDetailsFromAssertion(st *state.State, email string) (string, *osutil.AddUserOptions, error) {
+func getUserDetailsFromAssertion(o *overlord.Overlord, email string) (string, *osutil.AddUserOptions, error) {
```
Comment (stolowski) [inline daemon/api.go]: Same remark re passing overlord vs a simple ModelGetter interface or some such.

File: daemon/api.go (hunk 1 above)
Comment (pedronis) [reply daemon/api.go]: it's a fair remark, I kind of wondered the same but then the test change didn't get too much in the way (maybe we have too much setup there :) ),  but I can look at this again

Comment (bboozzoo) [reply daemon/api.go]: Could we pass `*asserts.Model` and state explicitly instead?

Comment (stolowski) [reply daemon/api.go]: The good thing about making the intent clear via an interface would be preventing of creating more (unintentional) coupling with overlord in the future, but that's maybe a theoretical concern. In any case, I suppose existing tests wouldn't be impacted (much) by such change, so maybe it's worth doing it.

File: overlord/devicestate/internal/state.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package internal (of devicestate) provides functions to access and
+// set the device state for use only by devicestate, for convenience they
+// are also exposed via devicestatetest for use in tests.
+package internal
```
Comment (bboozzoo) [inline overlord/devicestate/internal/state.go]: Looks like this was going to happen sooner or later.

File: daemon/api.go (hunk 1 above)
Comment (pedronis) [reply daemon/api.go]: Yes, passing `*asserts.Model` is probably the simplest, I might do that in one of the follow up though at this point.

**Pull request merged**